### PR TITLE
Admin, reports: improve customer addresses report

### LINKED ISSUE #11155: Improve Customer - Addresses Report
## Description

Please can we adjust the filter
- Add date filters like in Orders & Fulfilments reports 
- Change name of 'Distributor' filter to 'Hub'
- Remove the 'Supplier' filter

Please can we add the following columns to the end of the Customers - ~Mailing List~ Address report:
- Orders Count: number of orders for the customer, where filters are true
- Orders Total: total amount spent at the shop for the customer, where filters are true
- Last order: date of the last order, where filters are true.

Additionally can we rename this report to 'Customers'.

As a final step on this issue please remove the Mailing List report.

Clearly this is more than a papercut :( However since 4 instances chose this, let's just do it. 

## Acceptance Criteria & Tests

1. Three additional columns - Orders Count, Orders Total, Last Order
2. All new columns should display based on filters. In the case that only one order matches the filters, count=1, total=order total and last order = order date. In the case that no orders match the filters for the customer, that customer is not shown.
3. Date filters are added
4. Supplier filter removed
5. Distributor filter changed to Hub
6. Report renamed to Customers
7. Mailing List report removed
8. Specs updated and added
9. Translations on all fields.

### PR DESCRIPTION
**Pull request by jibees**
#### What? Why?

- Closes #11155 

This PR improve (by removing filters, adding some other ones, adding columns) the Customers Addresses report and delete the Customers Mailing List one.

##### Only one report for Customers
Capture d’écran 2023-07-10 à 15 04 53

##### Change filters: add date, remove Supplier, change Distributor to Hub
Capture d’écran 2023-07-10 à 15 05 01

##### Add columns: "Total Number of Orders", "Total incl. tax ($)" and "Last completed order date"
Capture d’écran 2023-07-10 à 15 05 06

##### Populate data
Capture d’écran 2023-07-10 à 15 05 14

#### What should we test?
This is pretty well explained in the original issue actually. 

#### Release notes

Changelog Category: User facing changes

### CODE REVIEW

File: lib/reporting/reports/customers/addresses.rb
Comment (rioug) [inline lib/reporting/reports/customers/addresses.rb]: Maybe you should move this to it's on method, it should help with the Cyclimatic Complexity

**Review by rioug — APPROVED** [review]
Nice one, I left a suggestion to try fix Rubocop warning but it shouldn't block merging.

File: lib/reporting/reports/customers/addresses.rb
Comment (jibees) [reply lib/reporting/reports/customers/addresses.rb]: Yes! 8fea985f2d

**Comment by mkllnk** [comment]
Moving to In Dev to resolve conflicts.

**Comment by jibees** [comment]
> Moving to In Dev to resolve conflicts.

thanks

Pretty though to resolve, hope everything goes well!

File: lib/reporting/reports/customers/base.rb
Comment (mkllnk) [inline lib/reporting/reports/customers/base.rb:61]: Here are some ideas:

* `fetch` is easier to reach into an array.
* Variable names add to understanding the code.
* A guard statement avoids indentation.
* A guard statement puts the focus on the main logic of the method.
* Using a value range is a bit shorter and clearer while doing the same.

```suggestion
          min = params.fetch(:q, :completed_at_gt)
          max = params.fetch(:q, :completed_at_lt)

          return orders if min.blank? || max.blank?

          orders.where(completed_at: [min..max])
```

File: spec/lib/reports/customers_report_spec.rb
Comment (mkllnk) [inline spec/lib/reports/customers_report_spec.rb:239]: This may just be my personal preference but I like to keep the code further to the left:
```suggestion
              allow(subject).to receive(:params).and_return(
                q: {
                   completed_at_gt: 1.day.before(o2.completed_at),
                   completed_at_lt: 1.day.after(o2.completed_at)
                 }
               )
```

It's easier to add to and if something changes, like `subject` is renamed to `customers_report` then the indentation of the following lines don't have to be changed.

File: lib/reporting/reports/customers/addresses.rb
Comment (mkllnk) [inline lib/reporting/reports/customers/addresses.rb]: The question is: Do we accept that column definitions are complex in nature and we don't want to change it? Or is this something we would like to solve one day?

For the former, I would add an exception to our Rubocop styleguide config to exclude the `columns` method. And for the latter I would add this to the Rubocop todo file instead of declaring it here.

I wish we could simplify the code but it would involve a major refactor of the report. So I actually think that your declaration here is the best solution! It is something we want to solve one day but we are okay with the current code. And when we refactor one day, the statement will disappear with a refactor. :+1:

File: spec/lib/reports/customers_report_spec.rb
Comment (mkllnk) [inline spec/lib/reports/customers_report_spec.rb:68]: I would prefer plain text here.

**Review by mkllnk — APPROVED** [review]
Excellent! Thank you for the detailed commit history. :star2:

File: lib/reporting/reports/customers/base.rb
Comment (jibees) [reply lib/reporting/reports/customers/base.rb]: oh wooo. Much more readable. Thanks for that! 🙏

File: spec/lib/reports/customers_report_spec.rb
Comment (jibees) [reply spec/lib/reports/customers_report_spec.rb]: Thanks! Yes, this is more readable !

File: spec/lib/reports/customers_report_spec.rb
Comment (jibees) [reply spec/lib/reports/customers_report_spec.rb:68]: Not sure to understand exactly why?

Comment (mkllnk) [reply spec/lib/reports/customers_report_spec.rb:68]: > Not sure to understand exactly why?

We came to [this convention](https://github.com/openfoodfoundation/openfoodnetwork/wiki/Code-Conventions#prefer-plain-text-over-method-calls-in-expected-values) for several reasons:

The code `strftime("%Y-%m-%d")` is the same in production code and in the spec. If it contains a mistake, for example using the wrong letter, then we don't notice this in the spec. The reader may also be left wondering how this formatting looks like exactly. Plain text is easier to read.

Comment (jibees) [reply spec/lib/reports/customers_report_spec.rb]: Thanks! Modified ;)

**Review by rioug — APPROVED** [review]

**Review by mkllnk — APPROVED** [review]

**Comment by filipefurtad0** [comment]
Hey @jibees 

Before this PR:

image

image


After this PR:

image

image

Since the addresses report now is missing, refreshing the page renders an error 500 - this is ok :ok_hand: 
https://app.bugsnag.com/yaycode/openfoodnetwork-uk/errors/64ca884cabf0a70008dec83d?event_id=64ca884c00becfd886240000&i=sk&m=nw

Maybe it happens that a user is visiting the page during deployment but... I think we can live with this :heavy_check_mark: 

As a hub manager,
visited the customers report, and went through (some) acceptance criteria:

image


1.    Three additional columns - Orders Count, Orders Total, Last Order :heavy_check_mark: 
2.    All new columns should display based on filters. In the case that only one order matches the filters, count=1, total=order total and last order = order date. In the case that no orders match the filters for the customer, that customer is not shown.

I've tried applying a filter by Order Cycle, but got:
https://app.bugsnag.com/yaycode/openfoodnetwork-uk/errors/64ca8a66b007c30008dba4f7?event_id=64ca8a6600becd56d5e90000&i=sk&m=nw

Also tried the Distributor filter (as hub manager and superadmin), which triggered:

https://app.bugsnag.com/yaycode/openfoodnetwork-uk/errors/64ca8ad8c0cc550008346333?event_id=64ca8ad800becbe8805f0000&i=sk&m=nw

4.    Date filters are added

Tried applying the date filter, got:

https://app.bugsnag.com/yaycode/openfoodnetwork-uk/errors/64ca8b2cc0cc550008346568?event_id=64ca8b2c00bebc227e310000&i=sk&m=nw

The filter works, if we only set an end date; if we set both dates (a date range) then we get the error 500 above :point_up: 

5.    Supplier filter removed :heavy_check_mark: 
6.    Distributor filter changed to Hub :heavy_check_mark: 
7.    Report renamed to Customers :heavy_check_mark: 
8.    Mailing List report removed :heavy_check_mark: 
9.    Specs updated and added :heavy_check_mark: - but maybe these can be improved? Maybe to consider filtering by distributor?
10.    Translations on all fields. :heavy_check_mark:

I think this one is not ready for shipping just yet... Moved back to in dev.

**Comment by jibees** [comment]
@filipefurtad0 Thanks for testing! I've failed during a refactoring, but seems corrected by ad375a607a Back in `Code Review` then.

**Review by mkllnk — APPROVED** [review]

**Review by rioug — APPROVED** [review]

**Comment by drummer83** [comment]
Hi @jibees,
That's a great one! :+1: 
Here are my testing notes.

# Test results
## Acceptance Criteria & Tests
1. Three additional columns - Orders Count, Orders Total, Last Order :heavy_check_mark: 
2. All new columns should display based on filters. In the case that only one order matches the filters, count=1, total=order total and last order = order date. In the case that no orders match the filters for the customer, that customer is not shown. :heavy_check_mark: 
3. Date filters are added :heavy_check_mark: 
4. Supplier filter removed :heavy_check_mark: 
5. Distributor filter changed to Hub :heavy_check_mark: 
6. Report renamed to Customers :heavy_check_mark: 
7. Mailing List report removed :heavy_check_mark: 
8. Specs updated and added :heavy_check_mark: 
9. Translations on all fields (:heavy_check_mark: as far as I can tell :question:)

## Additional request
- Add zip code or post code to the report (https://github.com/openfoodfoundation/openfoodnetwork/issues/11155#issuecomment-1638742348) :x: -> not implemented

## Additional tests
- Test as super admin :heavy_check_mark: 
- Report exports show the same results as on screen :heavy_check_mark: 

## Comparing the report results with orders
- Customer Details are correct :heavy_check_mark: 
- Order details are correct :heavy_check_mark: 
image
image
image

# Notes
- Regression: #10706 is back
- Date rage is not remembered after generating the report (#10392)
- It's not possible to remove the 'Hub' or 'Order Cycle' filter once it was added (page reload is required)
- It's not possible to select multiple hubs or order cycles in the 'Hubs' filter (like e.g. on the Enterprise Fee Summary report) -> use this implementation instead?
- Are the strings for distributor and supplier still used anywhere?
image

# Conclusion

- The report is working without errors and the specification has been followed. :muscle: 
- There is one minor open question for the devs regarding the potentially unused translation strings (see Notes section above).
- There are two questions for @openfoodfoundation/train-drivers-product-owners:
  - Should we avoid the regression of #10706 (see Notes section above) :question: 
  - Should we adjust the selection of hubs and order cycles before releasing (see Notes section above) :question: 

We could merge this one but I would like to see what product says about this. Adding feedback needed label.

Thanks again!

**Comment by drummer83** [comment]
Adding a video to show the behaviour around hubs and order cycle selection:


https://github.com/openfoodfoundation/openfoodnetwork/assets/1790176/09939730-6bca-43b6-b446-a03b3d6daeb1

**Comment by RachL** [comment]
Hello! Thanks @drummer83 ! Yes we should definitely avoid the regression of 10706.
Around the removal of filters, and multiple hub selection, it doesn't seem to be introduced by this PR. It seems this is already in production if I'm not wrong and was a regression due to another work. We should open a dedicated issue for this.

**Comment by drummer83** [comment]
Thanks, moving back to In Dev then.

**Comment by jibees** [comment]
Actually there are shown, because yes, they are considered as different rows, because of their different Shipping Method. When hiding shipping method, you want to _merge_  them into one row? Is that right? 

Sorry if it's a pretty noob question.

**Comment by drummer83** [comment]
It's not a noob question at all, @jibees!

Yes, however this applies to all columns. So if the Phone number or First name or Last name are different but the column is hidden, then these should be merged as well.

I hope you can have a look at #10706 and check how this was solved there.

**Comment by jibees** [comment]
Thanks @drummer83 !

Actually, this is supposed to work, because I can see this `skip_duplicate_rows` in the code. And yes, it works because this is different rows: Total included Tax, and Last completed order Date aren't the same. But maybe this should be merged before (in the query retrieving the data itself). 

Humm ... that's not totally clear to me. Let's think about that ... 🤔

**Comment by RachL** [comment]
I understand that it would be clean to merge linked to hide/show columns but I think that it might be difficult to maintain

In these customer reports, we want to retrieve mo up-to-date info on customer. Instead of merging - would it be a good strategy to display the last info for the customer ID?

**Comment by drummer83** [comment]
> Instead of merging - would it be a good strategy to display the last info for the customer ID?

That's a different approach and it would lead to different results, but I think it would work. And it also makes sense. :+1:

**Comment by sigmundpetersen** [comment]
> What do you mean by IRL?

In Real Life 😄

**Comment by RachL** [comment]
@jibees in delivery circle notes it is mentioned that this PR is pending feedback. So my last comment is not doable? Instead of showing all customer records, showing the last one for each customer ID

**Comment by jibees** [comment]
ok, now I see!

Fix made by #11148 isn't working anymore because we add columns "Total Number of Orders", "Total incl. tax ($)" and "Last completed order date" which is different (and couldn't be therefore merged). The spec didn't catch that because I forgot to add columns in the specs (shouldn't we show all the columns by default in the spec as well?).

**Comment by jibees** [comment]
> Instead of showing all customer records, showing the last one for each customer ID

Back in code review, since commit 9d8abbc150 that group by customer ID has been added.

File: lib/reporting/reports/customers/base.rb
Comment (mkllnk) [inline lib/reporting/reports/customers/base.rb:14]: It's possible to delete a customer record and then the customer id is null. This would group all deleted customers under one even though they were separate people. We could group by either customer id or email. Maybe not 100% correct in the long term because email addresses can be re-used by other people but we could do this:
```suggestion
              customer_id: order.customer_id || order.email,
              hub_id: order.distributor_id,
```

I added the hub_id in case the customer_id is not present. Otherwise it doesn't hurt because each customer id has exactly one distributor id.

**Review by mkllnk — CHANGES_REQUESTED** [review]

File: lib/reporting/reports/customers/base.rb
Comment (jibees) [reply lib/reporting/reports/customers/base.rb]: Oh nice catch, thanks a lot. Makes sense!!

File: lib/reporting/reports/customers/base.rb
Comment (rioug) [inline lib/reporting/reports/customers/base.rb:82]: This should be using `last_completed_order` defined above

Comment (jibees) [reply lib/reporting/reports/customers/base.rb:82]: Absolutely, that's why I've created this one ; sorry.

**Comment by drummer83** [comment]
Hi @jibees,
Thanks for continuing to work on this! :muscle:

Displaying the original columns (without number of orders, total amount, last order date) of this report is working perfectly now. No duplicates at all:
Customer report ok

However when I display the 'Total numbers of orders' I start seeing duplicates again:
Customer report 1 duplicate

Same for 'Total incl. tax' and 'Last completed order date':
Customer report 2 duplicates

I can't really tell why this is happening.

⚠️ Let's align with @RachL first, before fixing this last bit! My understanding was that it would be best to only show the latest customer data for each customer_ID, which makes the 'remove duplicates' function unnecessary - because customer_ID is always unique.

Sorry for this back and forth, I think we need a clear statement for how to preceed.
Thanks! 🙏

**Comment by jibees** [comment]
@drummer83 after some thoughts with Rachel, I'd like to see the second screenshot with the `hub` column selected ; to my understanding, it seems it's two different customers (yes with same name, address, ... but different hub) and therefore, not really duplicate. 
Am I clear enough? Sorry if not!

**Comment by jibees** [comment]
(rebased onto master, due to conflicting file `en.yml`)

**Review by mkllnk — APPROVED** [review]

**Comment by drummer83** [comment]
@jibees @RachL Sorry for not getting back to this earlier. I tried to prove my theory that we still have duplicates here, but I failed big time!! 
I am sure about one thing: I had filtered for one special hub, so the hub is the same for all entries in my screenshot. 

Is it possible that we have duplicates because the bill_address_id can be different even though the bill address actually looks identical? Just an idea based on my desperate efforts to replicate the report results via an psql query.

**Comment by mkllnk** [comment]
> Is it possible that we have duplicates because the bill_address_id can be different even though the bill address actually looks identical? Just an idea based on my desperate efforts to replicate the report results via an psql query.

The bill address id is always different but we don't group by that anymore. So we shouldn't see duplicates because of that.

**Comment by drummer83** [comment]
@jibees @RachL I just found that the orders marked in red in this screenshot are NOT duplicates because one of the orders doesn't have a customer_id attached. It's empty - no idea how this can happen.

Customer report 2 duplicates

```
   id   |   number   | item_total | total |  state   | adjustment_total | user_id |         created_at         |         updated_at         |        completed_at        | bill_address_id | ship_address_id | payment_total | shipment_state | payment_state |             email             | special_instructions | distributor_id | order_cycle_id | currency | last_ip_address | customer_id | created_by_id | included_tax_total | additional_tax_total | note 
--------+------------+------------+-------+----------+------------------+---------+----------------------------+----------------------------+----------------------------+-----------------+-----------------+---------------+----------------+---------------+-------------------------------+----------------------+----------------+----------------+----------+-----------------+-------------+---------------+--------------------+----------------------+------
 535713 | R716072052 |       3.00 | 11.04 | complete |             8.04 |    4378 | 2022-12-28 06:12:23.127855 | 2023-01-25 01:10:41.742754 | 2023-01-25 01:10:41.617877 |          347638 |          347642 |          0.00 | pending        | balance_due   | maikel@openfoodnetwork.org.au |                      |           2507 |           3809 | AUD      | 182.239.217.209 |             |          4378 |               0.00 |                -0.80 | 
 536357 | R845463134 |       9.00 | 24.03 | complete |            15.03 |    4378 | 2023-01-25 01:10:41.765772 | 2023-03-10 03:01:01.944059 | 2023-03-10 03:01:01.837768 |          347645 |          347759 |         24.03 | ready          | paid          | maikel@openfoodnetwork.org.au |                      |           2507 |           3809 | AUD      | 182.239.199.182 |        7083 |          4378 |               0.00 |                -1.50 | 
(2 rows)
```

**Comment by jibees** [comment]
Capture d’écran 2023-09-22 à 11 55 49

This was a debug. On first columns we can see all the customer_id for each orders that are merged to generate the current row (`|orders| orders.map{ |o| o.customer_id || o.email }.join(', ')`). The second is the `last_completed_order_id`. 

In this screenshot, you can see that the customer id 22849 (which actually has as email rachel.arnould@coopcircuits.fr stored in DB) but its email used on the order 2325494 it is well `rachel.arnould+release@protonmail.com` which leads to a duplicates feeling.

```
select email from customers where id = 22849;
             email
--------------------------------
 rachel.arnould@coopcircuits.fr
(1 row)
```

```
select email from spree_orders where id = 2325494;
                 email
---------------------------------------
 rachel.arnould+release@protonmail.com
```

But this is two different customers.

**Comment by jibees** [comment]
@drummer83 Having customer_id null is a normal behaviour.

> It's possible to delete a customer record and then the customer id is null.

from https://github.com/openfoodfoundation/openfoodnetwork/pull/11185#discussion_r1319255200

**Comment by RachL** [comment]
So, the rule seems respected. There is still a question around: should the report display customers who have been deleted? I would say no, but maybe this needs broader validation. I will ask on slack. In the meantime I think we can merge here and test with prod data if we are heading in the right direction?

**Comment by RachL** [comment]
I've checked with JB, apparently I can merge "without waiting for requirements to be met".

**Comment by RachL** [comment]
Arf no the branch is again out-of-date... if another @openfoodfoundation/core-devs can have a look?

**Comment by filipefurtad0** [comment]
Hey @RachL ,
Should be ok to merge, as long as there are no conflicts. If we really want to be sure, we can just Update with rebase - I took the liberty and did just that. If it turns out green then we're safe to merge :-)

**Comment by RachL** [comment]
Ah yes forgot that this could work, thanks @filipefurtad0 - fingers crossed

**Comment by RachL** [comment]
we are good to go! pressing the red button :see_no_evil:

**Comment by jibees** [comment]
> again out-of-date

Out-of-date doesn't mean it will surely cause any particular issue. It's not like merge conflict, which blocks the issue to be merged. Will merge this one!

**Comment by drummer83** [comment]
Added the prod-test label and moved to Test Ready as a production test was proposed by Rachel.

**Comment by drummer83** [comment]
I don't see any duplicate issues on German production server. It's working as I would expect.

Probably there is some weird data on the staging server that confused me.

**Comment by RachL** [comment]
@drummer83 thanks for checking this out! I was late to the party but no sign of duplicates in FR prod either

**Pull request merged**
